### ISSUE TITLE
major regression in parseapi parallelism

### ISSUE THREAD

**Issue by mwkrentel** [dyninst/dyninst#737]
I'm seeing a major regression in the ParseAPI parallelism between
10.1.0 and today's master 20673aae1701.

So far, I've only run big, complicated program (hpcstruct) on nwchem,
a large (160 Meg) x86_64 static binary.  But I'm pretty sure the
problem is on the dyninst side, as in it works as expected with 10.1.0
and I get terrible parallelism with master.

Running hpcstruct on nwchem with dyninst 10.1.0 and 12 threads on
mic.cs, I get times like this.

```
$ hpcstruct  --time  -j 12  Files/nwchem   
file:  /home/krentel/demo/Files/nwchem
symtab threads: 1  parse: 12  struct: 12

init:         0.0 sec         0 meg       171 meg
symtab:       7.0 sec      1165 meg      1337 meg
parse:       52.8 sec      7608 meg      8945 meg
struct:      14.2 sec      1050 meg      9995 meg
total:       74.1 sec      9824 meg      9995 meg

num funcs: 42043
```

Change dyninst 10.1.0 to master, and I get this.

```
file:  /home/krentel/demo/Files/nwchem
symtab threads: 1  parse: 12  struct: 12

init:         0.0 sec         0 meg       171 meg
symtab:       7.8 sec      1271 meg      1442 meg
parse:       33.7 sec     10087 meg     11530 meg
struct:     125.2 sec      1332 meg     12862 meg
total:      166.6 sec     12691 meg     12862 meg
```

It looks like you're shifting some of the computation.  The time for
the parse step goes down about 1/3, but memory goes up about 1/3.
But the big change is the queries from hpcstruct (traverse loop tree
and blocks) goes up almost 10x.

Also, in one run, I got some error messages.

```
Instruction jl 0x222(%rip)
type should be either COND_TAKEN or COND_NOT_TAKEN, but it is 5
From f78109 to f7810f
Instruction jl 0x222(%rip)
type should be either COND_TAKEN or COND_NOT_TAKEN, but it is 5
From f78109 to f7810f
```

I ran hpcrun on this.  In the trace viewer, I can see what's
happening.  Remember some 3 years ago, when Bill Williams still worked
there and we were first adding parallelism?  There was parallelism,
but it spent all its time waiting on locks.  This was before Xiaozhu
cleaned this out.  That's how it looks now.  In the struct phase, it
now spends 80% of its time waiting on locks inside dyninst.

I could send you trace snapshots or a database.

This was running hpcstruct on x86_64 on nwchem.  At some point, I'll
see if it also happens with my unit test.  It might not, the unit
test mostly targets the symtab and parse phases and runs a simpler
struct phase.

I can also try bisecting between 10.1.0 and master.
So much to do ....

@mxz297

**Comment by mxz297** [comment]
@mwkrentel I think I know what caused the struct time to explode. There is no need to do bisect. Please try this branch (https://github.com/mxz297/dyninst/tree/get_func_opt).

**Comment by mwkrentel** [comment]
@mxz297   Ok, that seems to restore the parallelism and produce traces
like I expect.

```
file:  /home/krentel/demo/Files/nwchem
symtab threads: 1  parse: 12  struct: 12

init:         0.0 sec         0 meg       171 meg
symtab:       7.6 sec      1271 meg      1442 meg
parse:       34.2 sec     10999 meg     12442 meg
struct:      15.1 sec      1466 meg     13908 meg
total:       56.9 sec     13737 meg     13908 meg
```

Compared to 10.1.0, this is overall maybe 25% faster, but 35% more
memory.  Memory is becoming the limiting factor, so this is moving in
the wrong direction.

I'm still seeing debug messages, although they're shorter.

```
From f78109 to f7810f
From f78109 to f7810f
From f78109 to f7810f
From f78109 to f7810f
```

**Comment by mxz297** [comment]
Closed via #729
